pkg/kubectl/cmd/util/factory.go

```diff
@@ -68,9 +68,57 @@ import (
 	"k8s.io/kubernetes/pkg/watch"
 )
 
-const (
-	FlagMatchBinaryVersion = "match-server-version"
-)
+// ClientCacheFactory provides access to a ClientCache
+type ClientCacheFactory interface {
+	// ClientCache returns a cache of clients.
+	ClientCache() ClientCache
+}
+
+// DiscoveryClientFactory provides access to a CachedDiscoverInterface
+type DiscoveryClientFactory interface {
+	// DiscoveryClient returns a discovery client
+	DiscoveryClient() (discovery.CachedDiscoveryInterface, error)
```

Comment:
If all you actually need is a way to initialize a factory with a different one these, the whole change is a lot more palatable.  Accepting one gives you `Object` and `UnstructuredObject` which looks like it will fix most things a lot more cleanly.